### PR TITLE
[Dynamo][6/N] Build torch in graph function (including constant folding ones) list as trace rules

### PR DESCRIPTION
**Pull request by yanboliang**
The goal of this PR is to build torch in graph function allow list, which is the most critical part of the Dynamo tracing rules. What I propose here is to dump from the ```is_allowed``` heuristic and iterator from there.
With the in graph function allow list, we can wrap them as ```TorchInGraphFunctionVariable```, and then we can probably remove the ```is_allowed``` and ```allowed_functions.py```, since the other things (torch classes, torch object, torch non-in-graph functions, torch modules) can be treated as regular python classes, objects, modules, functions (inline or skip).

cc @albanD @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: test/dynamo/test_trace_rules.py [hunk 1]
```diff
@@ -317,8 +378,14 @@ def test_skipfiles_inlinelist(self):
                 f"{f} from skipfiles.FUNC_INLINELIST is not a python function, please check and correct it.",
             )
 
+    @unittest.skip("debugging")
```
Comment (yanboliang) [inline test/dynamo/test_trace_rules.py:381]: This test is failed right now, there are a few ops mismatch, which I guess is the importing order issue, under debugging now.

File: test/functorch/test_eager_transforms.py [hunk 2]
```diff
@@ -307,6 +307,7 @@ def foo(x):
         with self.assertRaisesRegex(RuntimeError, 'mutate a captured Tensor'):
             grad(foo)(x)
 
+    @skipIfTorchDynamo("Debugging")
```
Comment (yanboliang) [inline test/functorch/test_eager_transforms.py:310]: This is double grad issue, debugging now.

File: test/functorch/test_eager_transforms.py [hunk 3]
```diff
@@ -1042,6 +1043,7 @@ def f(y, x):
         grad(f)(y, x)
         grad(grad(f))(y, x)
 
+    @skipIfTorchDynamo("Debugging")
```
Comment (yanboliang) [inline test/functorch/test_eager_transforms.py:1046]: double grad doesn't work well, debugging now.

File: test/test_nestedtensor.py [hunk 4]
```diff
@@ -3084,7 +3084,7 @@ def _validate_nt(self, nt, tensor_list, device, dtype, requires_grad):
         self.assertEqual(nt.values().device, device)
         self.assertEqual(nt.offsets().device, device)
         self.assertEqual(nt.shape[0], batch_size)
-        self.assertTrue(isinstance(nt.shape[1], torch.SymInt))
+        self.assertTrue(isinstance(nt.shape[1], (int, torch.SymInt)))
```
Comment (yanboliang) [inline test/test_nestedtensor.py:3087]: Latent bug, because of a ```SymInt``` object's type can be ```int``` in some cases, check the following code: 
https://github.com/pytorch/pytorch/blob/7b99b3efb1009668e0b77e35e9a51206709e11e4/torch/_dynamo/variables/tensor.py#L821-L825

File: torch/_dynamo/trace_rules.py [hunk 5]
```diff
@@ -35,40 +40,2732 @@
 and trace_rules.lookup consolidation is done. Then the explicit listing of skip/inline objects have
 a higher priority, which can be used to override the skipfiles.check rules in some cases.
 """
-torch_name_rule_map = {
+manual_torch_name_rule_map = {
```
Comment (yanboliang) [inline torch/_dynamo/trace_rules.py:43]: We have two maps:
* ```manual_torch_name_rule_map```: manually added by PT developers.
* ```auto_torch_name_rule_map```: map generated by the heuristic in ```test/dynamo/test_trace_rules.py```, which was copied from ```allowed_functions.py```.

The functions in ```manual_torch_name_rule_map``` are functions with Dynamo implementation. If in the future, we hit cases that need Dynamo to support a new function, we probably should put them in this list.

File: torch/_dynamo/trace_rules.py [hunk 6]
```diff
@@ -35,40 +40,2732 @@
...
     d = dict()
     for k, v in torch_name_rule_map.items():
-        obj = load_object(k)
-        assert obj not in d
-        d[obj] = v
+        try:
+            obj = load_object(k)
+            d[obj] = v
+        except Exception:
+            pass
```
Comment (yanboliang) [inline torch/_dynamo/trace_rules.py:2724]: The try ... except is to handle no CUDA functions in CPU only env, probably I should have a better way. But let me check if the idea is in right direction in case wasting time on trivial things.

**Comment by yanboliang** [comment]
@jansel @ezyang I'm posting this PR to collect early feedback on the in graph function changes. There are many trivial works need to be done, but I'd like to check if this is the right direction in case wasting time.

**Comment by yanboliang** [comment]
BTW, I also have another PR #112932 which was running full CI tests, the only difference is I didn't dump the in graph function list in that PR. It seems CI is almost green except some distributed tests I have to check with distributed folks.

File: torch/_dynamo/allowed_functions.py [hunk 7]
```diff
@@ -168,23 +171,57 @@ def _is_allowed_module_prefix(obj):
...
-            "torch.distributed.fsdp.",
-            "torch.distributed._tensor.",
-            # Inline through the ActivationWrapper in
-            # torch.distributed.algorithms._checkpoint.checkpoint_wrapper. This
-            # nn module calls torch.utils.checkpoint internally. If Dynamo does
-            # not trace this, AOT Autograd will try to trace this and can cause
-            # issues observed in
-            # https://github.com/pytorch/pytorch/issues/108269
-            "torch.distributed.algorithms.",
+            "torch._C._autograd",
```
[the remark this answers is not in the document]
Comment (jansel) [reply torch/_dynamo/allowed_functions.py:180]: Could one just import the other?

File: torch/_dynamo/trace_rules.py [hunk 8]
```diff
@@ -35,40 +40,2732 @@
...
+    "torch.distributed.get_world_size": TorchInGraphFunctionVariable,
+    "torch.distributed._tensor.DTensor#from_local": TorchInGraphFunctionVariable,
+    "torch._utils.is_compiling": TorchInGraphFunctionVariable,
+    "torch.overrides.get_default_nowrap_functions": TorchInGraphFunctionVariable,
+    "torch.fx._symbolic_trace.is_fx_tracing": TorchInGraphFunctionVariable,
+    "torch._dynamo.external_utils.is_compiling": TorchInGraphFunctionVariable,
+}
+
+
+auto_torch_name_rule_map = {
```
Comment (jansel) [inline torch/_dynamo/trace_rules.py:66]: Is this auto-generated?  If so include a docstring with the needed commands to regenerate it.

File: torch/_dynamo/trace_rules.py (hunk 6 above)
Comment (jansel) [reply torch/_dynamo/trace_rules.py:2724]: Let's make the `except` case more narrow here.

**Comment by yanboliang** [comment]
```
Your PR is 3281 LOC which is more than the 2000 maximum
allowed within PyTorch infra. PLease make sure to split up
your PR into smaller pieces that can be reviewed.
If you think that this rule should not apply to your PR,
please contact @albanD or @seemethere.
```
@albanD Can you suggest how to fix the LOC failure?

**Comment by jansel** [comment]
You are looking for the `skip-pr-sanity-checks` label.

File: torch/_dynamo/trace_rules.py (hunk 8 above)
Comment (ezyang) [reply torch/_dynamo/trace_rules.py:66]: I think @yanboliang's intention was that we would initially generate this, but then as the library evolves, we would manually update the list (so this would be the source of truth).

Seeing this here, there are a few questions that pop immediately to mind. The biggest one is how we are going to keep this list up-to-date with `native_functions.yaml` changes. A very common thing is that someone adds a new operator to `native_functions.yaml`, and then, under the changes here, this map will have to be updated. This seems suboptimal. In particular, for operators that are C bindings into the dispatcher, I think there's a case to be made that we should *not* have to manually update them.

One non-problem is what to do about Python function implementations in PyTorch core. I think the intention here is that, when someone adds a new Python function, our default assumption should be that we will Dynamo trace into it, not allow it in the graph. I think it'd be OK to argue about this though!

File: test/dynamo/test_trace_rules.py [hunk 9]
```diff
@@ -327,6 +394,7 @@ def test_torch_name_rule_map(self):
         )
         x = generated_torch_name_rule_set - used_torch_name_rule_set
         y = used_torch_name_rule_set - generated_torch_name_rule_set
+        # breakpoint()
```
Comment (ezyang) [inline test/dynamo/test_trace_rules.py:397]: remove

File: test/dynamo/test_trace_rules.py [hunk 10]
```diff
@@ -169,6 +174,27 @@ def generate_allow_list():
     torch_object_ids = dict()
     torch_objects = set()
 
+    def heuristic_record_if_in_graph_function(obj, module, name):
```
Comment (ezyang) [inline test/dynamo/test_trace_rules.py:177]: I have no idea what this does, so I suggest doc'ing it

File: test/dynamo/test_trace_rules.py [hunk 11]
```diff
@@ -169,6 +174,27 @@ def generate_allow_list():
...
+        try:
+            if hasattr(obj, "__wrapped__") and obj is not torch.ops:
+                obj = obj.__wrapped__
+        except Exception:
```
Comment (ezyang) [inline test/dynamo/test_trace_rules.py:181]: this try...except doesn't seem like it should be necessary

File: test/functorch/test_eager_transforms.py (hunk 2 above)
Comment (ezyang) [reply test/functorch/test_eager_transforms.py:310]: Wasn't the plan for this change to be semantics preserving?

File: test/test_nestedtensor.py (hunk 4 above)
Comment (ezyang) [reply test/test_nestedtensor.py:3087]: Shouldn't you xfail this test under dynamo instead? In eager shape[1] is guaranteed to be SymInt

File: torch/_dynamo/variables/torch.py [hunk 12]
```diff
@@ -148,36 +148,45 @@ def _dynamo_overriden_transformers_eq(self, other):
     pass
 
 
-def torch_reconstruct(codegen, value):
```
Comment (ezyang) [inline torch/_dynamo/variables/torch.py:151]: How should I understand, at a high level, the changes in this file?

File: torch/_dynamo/variables/builder.py [hunk 13]
```diff
@@ -733,6 +733,8 @@ def index_source(key):
                 guards=make_guards(GuardBuilder.ID_MATCH),
             )
         elif trace_rules.lookup(value) is not None:
+            if is_user_defined_allowed(value):
+                self.tx.output.has_user_defined_allowed_in_graph = True
```
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:737]: What does this do?

File: torch/_dynamo/trace_rules.py [hunk 14]
```diff
@@ -35,40 +40,2732 @@
 and trace_rules.lookup consolidation is done. Then the explicit listing of skip/inline objects have
 a higher priority, which can be used to override the skipfiles.check rules in some cases.
 """
-torch_name_rule_map = {
+manual_torch_name_rule_map = {
+    "torch.default_generator#get_state": TorchInGraphFunctionVariable,
```
Comment (ezyang) [inline torch/_dynamo/trace_rules.py:44]: What does the # mean here?

File: test/dynamo/test_trace_rules.py [hunk 15]
```diff
@@ -178,23 +204,56 @@ def _is_allowed_module_prefix(obj):
...
+            "torch.nn.parallel",
+            "torch.nn.utils",
+            "torch.multiprocessing",
+            "torch.onnx",
+            "torch.overrides",
+            "torch.package",
+            "torch.profiler",
+            "torch.serialization",
+            "torch.storage",
+            "torch.utils",
```
[the remark this answers is not in the document]
Comment (yanboliang) [reply test/dynamo/test_trace_rules.py:256]: I means using my judgement, there is no, and CI says no. Yes, I added them manually, to generate a smaller (and more accurate) in_graph_function list.

File: test/dynamo/test_trace_rules.py (hunk 1 above)
Comment (yanboliang) [reply test/dynamo/test_trace_rules.py:381]: Had fixed this issue, will update later.

File: test/functorch/test_eager_transforms.py (hunk 2 above)
Comment (yanboliang) [reply test/functorch/test_eager_transforms.py:310]: No, because we updated the ```is_allowed``` generating rule, though I think it doesn't matter. For this particular issue, I suspect it comes from the refactor in ```torch.py```, I have to figure out the root cause at least.

File: test/test_nestedtensor.py (hunk 4 above)
Comment (yanboliang) [reply test/test_nestedtensor.py:3087]: SG!

File: torch/_dynamo/allowed_functions.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (yanboliang) [reply torch/_dynamo/allowed_functions.py:180]: Will do!

File: torch/_dynamo/trace_rules.py (hunk 6 above)
Comment (yanboliang) [reply torch/_dynamo/trace_rules.py:2724]: Yes

File: torch/_dynamo/trace_rules.py (hunk 8 above)
Comment (yanboliang) [reply torch/_dynamo/trace_rules.py:66]: The ```auto_torch_name_rule_map``` is generated by the heuristic defined in ```test_trace_rules.py```, we will iterate from here, e.g, we have to remove the functions which should not be here, and move them to ```test_trace_rules.py:ignored_torch_name_rule_set``` to make the whole set on par with the set generated by the heuristic.
I think the ```test_trace_rules.py:test_torch_name_rule_map``` helped us to monitor changes happened inside of torch though it's not 100% guaranteed, but it's the best effort we can have.
Regarding to changes in ```native_functions.yaml```, can I say all functions in ```native_functions.yaml``` are in_graph_functions? If yes, probably I can add another rules to automatically update this map according the changes in ```native_functions.yaml```.

File: test/dynamo/test_trace_rules.py (hunk 10 above)
Comment (yanboliang) [reply test/dynamo/test_trace_rules.py:177]: Will do

File: torch/_dynamo/variables/torch.py (hunk 12 above)
Comment (yanboliang) [reply torch/_dynamo/variables/torch.py:151]: I refactor out a ```BaseTorchVariable``` as the base class for all torch variables. 
* ```TorchCtxManagerClassVariable```: torch context manager classes implemented by Dynamo;
* ```TorchInGraphFunctionVariable```: in graph functions (including constant folding functions and Dynamo implemented functions)
* ```TorchVariable```: The rest ```is_allowed``` torch classes/objects/functions. I'll further refactor this, and probably I'll remove this variable to in favor of ```UserDefineClassVariable```, ```PythonModuleVariable```, ```UserFunctionVariables``` for inlining ones, ```SkipFilesVariable``` for skipped ones.

File: torch/_dynamo/variables/builder.py (hunk 13 above)
Comment (yanboliang) [reply torch/_dynamo/variables/builder.py:737]: We added this tag every time we are trying to wrap a user defined allowed, so this is needed. But probably I can move this inside of the lookup function.

File: torch/_dynamo/trace_rules.py (hunk 14 above)
Comment (yanboliang) [reply torch/_dynamo/trace_rules.py:44]: ```get_state``` is a method rather than a function, so we need different identifier telling ```load_object``` to properly load it. This is rare cases for in graph functions.

File: test/dynamo/test_trace_rules.py (hunk 11 above)
Comment (yanboliang) [reply test/dynamo/test_trace_rules.py:181]: CPU only platform loading CUDA functions would throw ```AttributeError```, I'll update the exception to ```AttributeError```.

File: test/functorch/test_eager_transforms.py (hunk 2 above)
Comment (yanboliang) [reply test/functorch/test_eager_transforms.py]: Resolved.

**Review by albanD — COMMENTED** [review]
For automatically generated changes it is ok to skip the check as Jason mentioned.

File: torch/_dynamo/trace_rules.py (hunk 8 above)
Comment (ezyang) [reply torch/_dynamo/trace_rules.py:69]: Yes, everything `native_functions.yaml` is fine.

This is a little more complicated, though, because there is sometimes a nontrivial mapping from what is in native_functions.yaml and what actually pops up in the Python binding. This could easily end up being an entire PR's worth of work on its own.

File: test/dynamo/test_trace_rules.py (hunk 11 above)
Comment (ezyang) [reply test/dynamo/test_trace_rules.py]: What exactly is obj here, in this case?

File: torch/_dynamo/variables/builder.py (hunk 13 above)
Comment (ezyang) [reply torch/_dynamo/variables/builder.py:737]: So are you saying this is preexisting code? But if so, I would have expected that there would be a preexisting site you are deleting, but this is only added?

File: test/dynamo/test_trace_rules.py (hunk 11 above)
Comment (yanboliang) [reply test/dynamo/test_trace_rules.py]: My last comment is incorrect, I thought this is an exception at another place. This is used to record a decorated function, e.g., ```functools.lru_cache```. We should get the original function because the decorated object's type is not function.
try...except is not necessary any more, I'll remove it.

File: torch/_dynamo/variables/builder.py (hunk 13 above)
Comment (yanboliang) [reply torch/_dynamo/variables/builder.py:737]: Good catch! This is preexisting code. This was original at 
https://github.com/pytorch/pytorch/blob/d261687d5f56ac8148fab2567cf1fa6dd5264def/torch/_dynamo/variables/builder.py#L695-L701

Since I moved the in_graph_function wrapping out of ```is_allowed```, so I should remove them from the old place.

File: test/dynamo/test_trace_rules.py (hunk 11 above)
Comment (yanboliang) [reply test/dynamo/test_trace_rules.py]: The obj is any obj under torch, we will unwrap to get the original object. And the next few lines to check if this obj is a function/method type and record it. This is used to generate the in_graph_function list based on the heuristic.

**Comment by ezyang** [comment]
some notes from our 1:1

* need to review all new function changes
* @albanD public function bindings list
* ideal end state? 
* merge conflicts / having to wait for CI to tell you something
* by default, we will start inlining all Python functions in dynamo
* use `__code__` to figure out if something is C binding or not / pierce through wrappers; or traverse `_C` to find all of C bindings
* various scenario: developers add a new C binding function (three possibilities: implement in dynamo, graph break it, allow in graph)

**Comment by yanboliang** [comment]
I'll split this change into several PRs based on the discussion with @ezyang:
* PR1: Refactor out ```TorchInGraphFunction``` and generate the list on the fly with the updated heuristic.
* PR2: Dump the in graph function list(~2600 ops) and adding unit tests.
  * Going forward, we assume all new functions are not in graph functions except functions in ```native_functions.yaml``` and some C bindings. So we don't need to monitor changes in every submodule.
  * Update the list automatically if adding/removing functions in ```native_functions.yaml```.
  * Update the unit test to only monitor C bindings changes. 
  * Make the list sorted alphabetically. 
* PR3: Make the in graph function list more accurate. 
* PR4: Documentation.
  * As a developer, how to update the in graph function list if they adding a new C binding.
  * How does the new ```trace_rule.lookup``` API works after consolidating in graph, inline and skip rule check? E.g, first check ```torch_name_rule_map```, then check inline or skip rule.
